### PR TITLE
MAT-9966: Add endpoints for valueset and codesystem

### PR DESCRIPTION
**Pull request by chubert-sb**
## Terminology Service PR

Jira Ticket: [MAT-9966](https://jira.cms.gov/browse/MAT-9966)
(Optional) Related Tickets:

### All Submissions
* [ ] This PR has the JIRA linked.
* [ ] Required tests are included.
* [ ] No extemporaneous files are included (i.e Complied files or testing results).
* [ ] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/gov/cms/madie/terminology/service/ValueSetExpansionService.java [hunk 1]
```diff
@@ -244,4 +250,36 @@ private Optional<ValueSet> parseExpansionResponse(String rawResponse) {
...
+                        String.format("ValueSet not found for url: %s version: %s", url, version)));
+    log.info(
+        "Successfully retrieved ValueSet with url: [{}] version: [{}] valueSet: [{}]",
+        url,
+        version,
+        valueSet);
+    return valueSet;
+  }
+
+  // In ValueSetExpansionService.java
```
Comment (jkotanchik-SB) [inline src/main/java/gov/cms/madie/terminology/service/ValueSetExpansionService.java]: Comment not particularly helpful.

Comment (chubert-sb) [reply src/main/java/gov/cms/madie/terminology/service/ValueSetExpansionService.java]: fixed it

**Review by jkotanchik-SB — DISMISSED** [review]

File: src/test/java/gov/cms/madie/terminology/config/ApiKeyFilterTest.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        ServletOutputStream servletOutputStream =
+                new ServletOutputStream() {
+                    @Override
+                    public boolean isReady() {
+                        return true;
+                    }
+
+                    @Override
+                    public void setWriteListener(WriteListener writeListener) {
```
Comment (jkotanchik-SB) [inline src/test/java/gov/cms/madie/terminology/config/ApiKeyFilterTest.java]: Codacy grumps at this empty method. If it cannot be removed, add a comment clarifying it is blank on purpose. Just like textbooks!

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    // Validate URL format
+    if (!isValidUrl(url)) {
+      throw new IllegalArgumentException("Invalid URL format");
+    }
+
+    return ResponseEntity.ok(vses.getCodeSystem(url, count));
+  }
+
+  private boolean isValidUrl(String url) {
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: Nitpicking: Can the method name be more specific?

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+    // Validate URL format
+    if (!isValidUrl(url)) {
+      throw new IllegalArgumentException("Invalid URL format");
+    }
+
+    return ResponseEntity.ok(vses.getValueSet(url, version).getValueSet());
+  }
+
+  @GetMapping("/CodeSystem")
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: We have been following the Standard convention for API End points, lowercase with hypens, and also plural nouns - since it is a list of code systems ( Ex: code-systems )

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+@Slf4j
+@RequiredArgsConstructor
+public class ValueSetExpansionController {
+
+  private final ValueSetExpansionService vses;
+
+  @Value("${madie.allowed-hosts}")
+  private List<String> allowedHosts;
+
+  @GetMapping("/ValueSet")
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: Similar comment as above

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.web.bind.annotation.*;
+
+import java.net.URL;
+import java.util.List;
+
+@RestController
+@RequestMapping(path = "/terminology")
+@Slf4j
+@RequiredArgsConstructor
+public class ValueSetExpansionController {
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: It could be me, but we have quite a few number of Value Set expansion end points, So it is super confusing to what is the difference between these multiple VSExpansions. To make thing organized, do you think we can rename the Controller name ? especially when this controller also has Fetching codeSystem end point and not just valueSets

Comment (RohitKandimalla) [reply src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: I see that the name came from "ValueSetExpansion*** vses" so it is tricky

File: src/main/resources/application.yml [hunk 7]
```diff
@@ -8,6 +8,9 @@ madie:
   allowedApi: http://localhost:9000
   user-service:
     base-url: ${USER_SERVICE_URL:http://localhost:8088/api}
+  api-key: ${TERMINOLOGY_API_KEY:0a51991c}
+  api-key-header: ${API_KEY_HEADER:api-key}
```
Comment (RohitKandimalla) [inline src/main/resources/application.yml:12]: Why do we need api-key ? Aren't we moving away from api-key in favor of savyint roles ?

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    // Validate URL format
+    if (!isValidUrl(url)) {
+      throw new IllegalArgumentException("Invalid URL format");
+    }
+
+    return ResponseEntity.ok(vses.getValueSet(url, version).getValueSet());
+  }
+
+  @GetMapping("/CodeSystem")
+  public ResponseEntity<List<CodeSystem>> expandCodeSystem(
```
Comment (RohitKandimalla) [inline src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: Technically I am not in favor of this title, since we are not expanding code systems yet, We are retrieving all versions of code systems. I believe the returned List of CodeSystem doesn't have any code concepts.

File: src/main/resources/application.yml (hunk 7 above)
Comment (jkotanchik-SB) [reply src/main/resources/application.yml:12]: Different context. Savyint will provide user roles. This api-key is for service-to-service communication.

Comment (RohitKandimalla) [reply src/main/resources/application.yml:12]: Got it

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java (hunk 3 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: I'm terrible at naming. Maybe isAllowableUrl better?

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java (hunk 4 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: I was just following the AC's here. I updated in code and I'll make a note in JIRA

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java (hunk 6 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: Yeah not sure what would be best here because of that. I tried VSESController? I'm not ecstatic about that name either. Let me know if you have a better name.

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java (hunk 8 above)
Comment (chubert-sb) [reply src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: I agree. I updated to something that should be better.

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java (hunk 3 above)
Comment (RohitKandimalla) [reply src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: How about isValidFhirTerminologyUrl ?

File: src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java (hunk 6 above)
Comment (RohitKandimalla) [reply src/main/java/gov/cms/madie/terminology/controller/ValueSetExpansionController.java]: We can leave it for now I guess, but we need a better name for VSES, We are not dealing with just ValueSets anymore. I am 50% lost on the flow anyways. I will mark it resolved, we can keep an eye on the naming and change it in future.

**Review by RohitKandimalla — APPROVED** [review]
